s3migrate/shared.py

```diff
@@ -0,0 +1,397 @@
+"""Tools for migrating Mogile content to S3."""
+
+import base64
+import hashlib
+import json
+import os
+import random
+import shutil
+import tempfile
+import threading
+from queue import Empty, Queue
+
+import dj_database_url
+import pymysql
+import requests
+from botocore.exceptions import ClientError
+
+BUFSIZE = 16*1024*1024  # 16 MiB
+
+
+def make_bucket_map(buckets):
+    """Construct a hash from a buckets source string of the form a:b,c:d."""
+    retval = {}
+    for bucket in buckets.split(','):
+        source, target = bucket.split(':')
+        retval[source] = target
+    return retval
+
+
+def hash_fileobj(fileobj, hasher):
+    """Efficiently hash a file object using the provided hasher."""
+    blocksize = 65536
+    fileobj.seek(0)
+    block = fileobj.read(blocksize)
+    while block:
+        hasher.update(block)
+        block = fileobj.read(blocksize)
+    fileobj.seek(0)
+    return hasher
+
+
+def md5_fileobj_hex(fileobj):
+    """Efficiently calculate the MD5 sum for a file object.
+
+    Returns a hex string.
+    """
+    return hash_fileobj(fileobj, hashlib.md5()).hexdigest()
+
+
+def md5_fileobj_b64(fileobj):
+    """Efficiently calculate the MD5 sum for a file object.
+
+    Returns a Base64 encoded string.
+    """
+    return base64.b64encode(
+        hash_fileobj(fileobj, hashlib.md5()).digest()
+    ).decode("utf-8")
+
+
+def sha1_fileobj_hex(fileobj):
+    """Efficiently calculate the SHA1 sum for a file object.
+
+    Returns a hex string.
+    """
+    return hash_fileobj(fileobj, hashlib.sha1()).hexdigest()
+
+
+def sha1_fileobj_b64(fileobj):
+    """Efficiently calculate the SHA1 sum for a file object.
+
+    Returns a Base64 encoded string.
+    """
+    return base64.b64encode(
+        hash_fileobj(fileobj, hashlib.sha1()).digest()
+    ).decode("utf-8")
+
+
+class MogileFile():
+    """Represents a file stored in mogile."""
+
+    def __init__(self, fid: int, dkey: str, length: int):
```

Comment:
ayyy, look at that type hinting

Reply:
I was a little lazy but I did do some of type stuff.